internal/repo/directory.go

```diff
@@ -42,29 +41,31 @@ func getParentDir(dir string) string {
 
 // UpsertParentDirs updates all parent directories of an object name in one transaction
 func (d *Directory) UpsertParentDirs(bucket string, objName string, newSize int64, newCount int64) error {
-	// getUpsertQuery returns a full query string to insert or update a directory
-	getUpsertQuery := func(bucket string, name string, size int64, count int64) string {
-		parentDir := getParentDir(name)
-
-		return fmt.Sprintf(`
+	query := `
 			INSERT INTO directory (bucket, name, size, count, parent)
-			VALUES ('%s','%s', %d, %d,'%s')
+			VALUES ($1, $2, $3, $4, $5)
 			ON CONFLICT(bucket, name)
 			DO UPDATE
-			SET size = size + %d,
-				count = count + %d;
-			`, bucket, name, size, count, parentDir, size, count)
-	}
+			SET size = size + $3,
+				count = count + $4;
+	`
 
 	if len(bucket) == 0 || len(objName) == 0 {
 		return errors.New("bucket or name argument is empty")
 	}
 
 	dirName := getParentDir(objName)
-	query := "BEGIN IMMEDIATE;"
+
+	tx, err := d.DB.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
```

Comment:
should we rollback transaction always?

Reply:
If transaction succeeds at committing, `tx.Rollback()` will be a [no-op](https://go.dev/doc/database/execute-transactions#:~:text=making%20the%20deferred%20rollback%20call%20a%20no%2Dop). This way we don't have to call for rollback on the different error checks

Reply:
please add a comment before merging